### PR TITLE
H2: support embedding the DB engine in native images

### PR DESCRIPTION
**Pull request by Sanne**
This introduces the ability to embed the whole H2 engine within native images.

Reviewing the integration tests is not so interesting: it's a subset copy of the existing ones we had in `jpa-h2`: a subset because many have been added there which test particular proxy behavious of Hibernate ORM, I don't think it would be useful to duplicate those but rather keeping it to "does H2 work" - so I kept the basic CRUD tests only.

A new Maven module was necessary for the integration tests as I needed:
 - different configuration to use H2 in embedded mode.
 - different dependencies: avoid the dev-services, and also avoid the `quarkus-test-h2` helpers.

The good news: all Substitutions removed!

# New builditem: `NativeImageEnableModule`
I needed to introduce a new general-purpose build item though; the H2 code refers to optional symbols in the `jdk.net` module, which isn't activated by default; apparently GraalVM introduced an `--add-modules` argument which addresses such needs. A bit puzzling that we never needed this so far, it seems others have been living with this limitation with other (more complex?) workarounds.

## Limitations
Integration tests are not very extensive: the embedded engine seems to work fine in its "in memory" configuration (as tested); but disk storage and clustering capabilities are untested; I consider this beyond current scope, but it warrants being considered experimental.

### CODE REVIEW

**Comment by Sanne** [comment]
Rebased to incorporate @gastaldi 's suggestion and cleanup history of changesets.

Thanks @gastaldi !

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java [hunk 1]
```diff
@@ -859,6 +870,11 @@ public NativeImageInvokerInfo build() {
                         }
                     }
                 }
+                if (enableModules != null && enableModules.size() > 0) {
+                    String modules = enableModules.stream().map(NativeImageEnableModule::getModuleName).distinct().sorted()
+                            .collect(Collectors.joining(", "));
```
[the remark this answers is not in the document]
Comment (Sanne) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:875]: yes! :100:

**Comment by Sanne** [comment]
rebased to incorporate @Ladicek 's fix too. Thanks!

**Review by gastaldi — APPROVED** [review]

File: docs/src/main/asciidoc/datasource.adoc [hunk 2]
```diff
@@ -484,9 +484,12 @@ Our suggestion is to use the real database you intend to use in production; cont
 good reasons to also want the ability to run quick integration tests using the JVM powered databases,
 so this is possible as well.
 
-It is important to remember that when configuring H2 (or Derby) to use the embedded engine,
+It is important to remember that when configuring Derby to use the embedded engine,
 this will work as usual in JVM mode but such an application will not compile into a native executable, as the Quarkus extensions only cover for making the JDBC client code compatible with the native compilation step: embedding the whole database engine into a native executable is currently not implemented.
 
+Compiling H2 into a native-image is now possible (it didn't use to be) but this is considered experimental; even if this technically possible
```
Comment (gsmet) [inline docs/src/main/asciidoc/datasource.adoc:490]:
```suggestion
Compiling H2 into a native-image is now possible (it didn't use to be) but this is considered experimental; even if this is technically possible,
```

File: docs/src/main/asciidoc/datasource.adoc [hunk 3]
```diff
@@ -484,9 +484,12 @@ Our suggestion is to use the real database you intend to use in production; cont
...
+we would still recommend to consider carefully why you'd want to embed H2 within your native-image.
```
Comment (gsmet) [inline docs/src/main/asciidoc/datasource.adoc:491]:
```suggestion
we would still recommend to consider carefully why you would want to embed H2 within your native-image.
```
(I thin the doc team would rewrite that)

**Comment by Sanne** [comment]
Thanks @gsmet , applied your suggestions as well.

**Comment by Sanne** [comment]
I'll merge it to spare  CI - it was green before , and only rebased to squash the docs rewording.

**Pull request merged**